src/cdomains/intDomain.ml

```diff
@@ -779,6 +804,69 @@ struct
     in
     QCheck.(set_shrink shrink @@ set_print show @@ map (*~rev:BatOption.get*) (of_interval ik) pair_arb)
   let relift x = x
+
+  let abs x = let r = Ints_t.neg x
+   in if Ints_t.compare x r < 0 then r else x
+
+  let modulo n k =
+    let result = Ints_t.rem n k in
+    if Ints_t.compare result Ints_t.zero >= 0 then result
+    else Ints_t.add result  k
+
+  let refine_with_congruence ik (intv : t) (cong : (int_t * int_t ) option) : t =
+    match intv, cong with
+    | Some (x, y), Some (c, m) ->
+        if (m = Ints_t.zero) && (Ints_t.compare c x < 0 || Ints_t.compare c y > 0) then None
+        else if m = Ints_t.zero then Some (c, c)
+        else let rcx =
+             if Ints_t.equal x (min_int ik) then x else
+             Ints_t.add x (modulo (Ints_t.sub c x) (abs(m))) in
+             let lcy =
+              if Ints_t.equal y (max_int ik) then y else
+              Ints_t.sub y (modulo (Ints_t.sub y c) (abs(m))) in
+             if Ints_t.compare rcx lcy > 0 then None
+             else if Ints_t.compare rcx lcy = 0 then norm ik @@ Some (rcx, rcx)
+             else norm ik @@ Some (rcx, lcy)
+    | _ -> intv
+
+  let refine_with_congruence ik x y =
+    let refn = refine_with_congruence ik x y in
+    if M.tracing then M.trace "refine" "int_refine_with_congruence %a %a -> %a\n" pretty x pretty y pretty refn;
+    refn
+
+  let refine_with_interval ik a b =
+    match b with
+    | None -> a
```

Comment:
Why is this `a` if `b` is `bot`? Do we have the same problem as with the lists here?

Reply:
It would make more sense to return `bot` instead. The reason why I put `a` there was because I incorrectly assumed that `refine_with_interval `and `refine_with_congruence` would also be called if interval/congruence analysis were disabled and parameter b would therefore become `bot`. Instead this is filtered by the `maybe` function. I checked the code again quickly and saw that this problem occurs in a few domains. I guess we can simply return `bot` everywhere where `refine_with_interval` /`refine_with_congruence` are called with param b = `bot`.